pkg/utils/has/controller.go

```diff
@@ -486,30 +486,105 @@ func (h *SuiteController) GetComponentService(componentName string, componentNam
 	return service, nil
 }
 
-func (h *SuiteController) WaitForComponentPipelineToBeFinished(c *common.SuiteController, componentName, applicationName, componentNamespace, sha string) error {
-	return wait.PollImmediate(20*time.Second, 30*time.Minute, func() (done bool, err error) {
-		pipelineRun, err := h.GetComponentPipelineRun(componentName, applicationName, componentNamespace, sha)
+func (h *SuiteController) WaitForComponentPipelineToBeFinished(component *appservice.Component, sha string, maxRetries int) error {
+	attempts := 1
+	app := component.Spec.Application
+	var pr *v1beta1.PipelineRun
 
-		if err != nil {
-			GinkgoWriter.Println("PipelineRun has not been created yet")
-			return false, nil
-		}
+	for {
+		err := wait.PollImmediate(20*time.Second, 30*time.Minute, func() (done bool, err error) {
+			pr, err = h.GetComponentPipelineRun(component.GetName(), app, component.GetNamespace(), sha)
+
+			if err != nil {
+				GinkgoWriter.Println("PipelineRun has not been created yet")
+				return false, nil
+			}
 
-		for _, condition := range pipelineRun.Status.Conditions {
-			GinkgoWriter.Printf("PipelineRun %s reason: %s\n", pipelineRun.Name, condition.Reason)
+			GinkgoWriter.Printf("PipelineRun %s reason: %s\n", pr.Name, pr.GetStatusCondition().GetCondition(apis.ConditionSucceeded).GetReason())
 
-			if !pipelineRun.IsDone() {
+			if !pr.IsDone() {
 				return false, nil
 			}
 
-			if pipelineRun.GetStatusCondition().GetCondition(apis.ConditionSucceeded).IsTrue() {
+			if pr.GetStatusCondition().GetCondition(apis.ConditionSucceeded).IsTrue() {
 				return true, nil
 			} else {
-				return false, fmt.Errorf(tekton.GetFailedPipelineRunLogs(c, pipelineRun))
+				err = tekton.StorePipelineRun(pr, h.KubeRest(), h.KubeInterface())
+				if err != nil {
+					GinkgoWriter.Printf("failed to store pipelineRun: %s\n", err.Error())
+				}
+				return false, fmt.Errorf(tekton.GetFailedPipelineRunLogs(h.KubeRest(), h.KubeInterface(), pr))
+			}
+		})
+
+		if err != nil {
```

Comment:
Design nitpick/question:

Conceptually, to me, this feels awkward to embed this retry functionality in our HAS controller? I would have thought this would be encapsulated in a different function that could decorate a function that is waiting for a pipelinerun? But maybe this is ok, I think I'm still trying to make sense of our Domain Concepts within the framework

Reply:
I tried to use decorators, but the code didn't look better. I didn't find any advantage of using the decorator in this case - it could perhaps make sense if our test framework would be designed differently. I'm happy to revisit this in a future though, maybe I chose a wrong approach.

So what I did instead is that I moved the retry logic to the [other method of has controller](https://github.com/redhat-appstudio/e2e-tests/pull/535/files#diff-80ebb5c0d3940057830cec3ae4ff0e5592619cf4a3d2087e697ebf146e9101f3R659) - reason for that is that the method itself contains other has controller method calls, so it's easier to handle it inside the HAS controller method